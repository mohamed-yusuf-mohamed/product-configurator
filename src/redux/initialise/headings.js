const headings = {
  summary: {
    "page-count": "Pages",
    "cover-colour": "Colour",
    "paper-type": "Layout",
    price: "Price",
  },
  options: {
    "cover-colour": "Choose your colour",
    "paper-type": "Choose your page layout",
  },
}
export default headings
