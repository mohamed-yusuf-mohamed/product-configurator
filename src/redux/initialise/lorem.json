"Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla quis justo eget augue mattis volutpat. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Proin lectus diam, lobortis ut commodo in, accumsan non orci. Quisque ut tortor ut ligula volutpat imperdiet a eu mauris. Nulla ullamcorper nunc est, vitae ornare tellus pulvinar efficitur. Fusce justo neque, dapibus in gravida nec, eleifend sed velit. Integer vehicula sodales felis, lacinia pretium nulla. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Vivamus vitae malesuada nibh. Cras ultrices libero feugiat ullamcorper posuere. Maecenas elementum quam non dui vestibulum, ut tempus augue facilisis. Etiam non ipsum vitae purus eleifend egestas ac blandit turpis. Fusce mi massa, interdum sed mi et, malesuada ultricies diam."

