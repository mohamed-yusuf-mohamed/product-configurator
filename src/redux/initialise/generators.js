import lorem from "./lorem.json"
import headings from "./headings"

export const generateProducts = (data) => {
  return data.products.reduce(
    (prev, product) => ({
      ...prev,
      [product.id]: {
        id: product.id,
        title: product["product-label"],
        description: lorem,
        image: "./moo-product-image.jpeg",
        price: product.price,
        attributes: {
          ...product.attributes.reduce(
            (prev, attribute) => ({
              ...prev,
              [attribute.type]: attribute,
            }),
            {}
          ),
        },
      },
    }),
    {}
  )
}

export const generateOptions = (data) => {
  const products = data.products

  const attributes = Object.values(products).reduce((prev, product) => {
    return [...prev, ...product.attributes]
  }, [])

  const selectable = []
  for (const attribute of attributes) {
    if (attribute.selectable) selectable.push(attribute)
  }

  const values = selectable.reduce(
    (prev, attribute) => ({
      ...prev,
      [attribute.value]: attribute,
    }),
    {}
  )

  return Object.values(values).reduce(
    (prev, attribute) => ({
      ...prev,
      [attribute.type]: {
        type: attribute.type,
        label: headings.options[attribute.type],
        values: {
          ...prev[attribute.type]?.values,
          [attribute.value]: {
            value: attribute.value,
            label: attribute.label,
          },
        },
      },
    }),
    {}
  )
}
