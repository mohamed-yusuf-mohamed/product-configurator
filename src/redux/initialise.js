import {initialiseData} from "./actions"
import {store} from "./store"
import api from "../api"
import {generateProducts, generateOptions} from "./initialise/generators"
const initialise = async () => {
  const data = await api.getProducts()
  const products = generateProducts(data)
  const options = generateOptions(data)
  const defaultSelection = products[1]

  store.dispatch(initialiseData({
    products,
    options,
    selected: defaultSelection
  }))
}
export default initialise