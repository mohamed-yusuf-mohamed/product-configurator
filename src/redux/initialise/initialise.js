import {initialiseData} from "../actions"
import {store} from "../store"
import api from "../../api"
import {generateProducts, generateOptions} from "./generators"
const initialise = async () => {
  const data = await api.getProducts()
  const products = generateProducts(data)
  const options = generateOptions(data)
  const defaultSelection = products[1]

  store.dispatch(initialiseData({
    products,
    options,
    selected: defaultSelection
  }))
  
  return store
}
export default initialise