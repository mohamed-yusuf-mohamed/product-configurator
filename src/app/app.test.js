import { render, fireEvent } from '@testing-library/react';

// import {fetchData, handleInput, reset, load} from "./actions"
// import fetch from "jest-fetch-mock"
// import initialState from "./mocks/state.json"
import initialise from "../redux/initialise"
import App from "./index"
// import initialise from "../redux/initialise"
// import {store} from "./store"
// import { enableFetchMocks } from 'jest-fetch-mock'
// enableFetchMocks()





describe("app", () => {
  
  
  it("renders: matches snapshot", () => {
    initialise().then((store) => {
      const {getByTestId} = render(<App store={store}/>) 
      const component = getByTestId("product-configurator")
    
      expect(component).toMatchSnapshot()
    })

  })

  })


  