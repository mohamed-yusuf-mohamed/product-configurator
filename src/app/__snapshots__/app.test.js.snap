// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app renders: matches snapshot 1`] = `
<div
  class="sc-dkrFOg hxKPMK"
  data-testid="product-configurator"
>
  <div
    class="title"
  >
    <h1>
      MOO Hardcover Notebook
    </h1>
  </div>
  <div
    class="body"
  >
    <div
      class="image"
    >
      <img
        src="./moo-product-image.jpeg"
      />
    </div>
    <div
      class="configurator"
    >
      <div
        class="description"
      >
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nulla quis justo eget augue mattis volutpat. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Proin lectus diam, lobortis ut commodo in, accumsan non orci. Quisque ut tortor ut ligula volutpat imperdiet a eu mauris. Nulla ullamcorper nunc est, vitae ornare tellus pulvinar efficitur. Fusce justo neque, dapibus in gravida nec, eleifend sed velit. Integer vehicula sodales felis, lacinia pretium nulla. Vestibulum ante ipsum primis in faucibus orci luctus et ultrices posuere cubilia curae; Vivamus vitae malesuada nibh.
      </div>
      <div
        class="button-group"
        data-testid="cover-colour-options"
      >
        <p>
          Choose your colour
        </p>
        <button
          class="sc-gswNZR iwUBsT"
          data-testid="option-button"
        >
          Sunset
        </button>
        <button
          class="sc-gswNZR KNFML"
          data-testid="option-button"
        >
          Ocean
        </button>
      </div>
      <div
        class="button-group"
        data-testid="paper-type-options"
      >
        <p>
          Choose your page layout
        </p>
        <button
          class="sc-gswNZR iwUBsT"
          data-testid="option-button"
        >
          Dotted
        </button>
        <button
          class="sc-gswNZR KNFML"
          data-testid="option-button"
        >
          Lined
        </button>
      </div>
      <div
        class="sc-bcXHqe iPrJEm"
        data-testid="product-summary"
      >
        <table>
          <tbody>
            <tr>
              <th>
                Summary
              </th>
              <th
                class="hide"
              >
                Values
              </th>
            </tr>
            <tr>
              <td>
                Pages
              </td>
              <td>
                170
              </td>
            </tr>
            <tr>
              <td>
                Colour
              </td>
              <td>
                Sunset
              </td>
            </tr>
            <tr>
              <td>
                Layout
              </td>
              <td>
                Dotted
              </td>
            </tr>
            <tr>
              <td>
                Price
              </td>
              <td>
                Â£17.00
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </div>
</div>
`;
